sys/phydat: added phydat to JSON converter

haukepetersen opened the pull request:
~~rebased on #7263~~

This PR adds means for converting a `phydat_t` structure in a simple JSON string. The function is build as simple as possible, no means for templating or similar... I would imagine, that we introduce more sophisticated converters at some point when needed, e.g. something like `phydat_t` to IPSO smart object and so on. 

So this PR is rather a starting point to see what we can do in this direction, but it should already be useful in demos etc...

haukepetersen commented:
By the way, next step: phydat to CBOR :-)

aabadie commented on sys/include/phydat.h:
```diff
@@ -157,6 +158,23 @@ void phydat_dump(phydat_t *data, uint8_t dim);
+ * @brief   Return a string representation for every unit, including none
+ *          physical units like 'bool' or 'time'
+ *
+ * This function is useful when converting phydat_t structures to none-binary
```
non-binary

aabadie commented on sys/include/phydat.h:
```diff
@@ -157,6 +158,23 @@ void phydat_dump(phydat_t *data, uint8_t dim);
 const char *phydat_unit_to_str(uint8_t unit);
 
 /**
+ * @brief   Return a string representation for every unit, including none
```
non

aabadie commented on sys/include/phydat.h:
```diff
@@ -157,6 +158,23 @@ void phydat_dump(phydat_t *data, uint8_t dim);
+ * This function is useful when converting phydat_t structures to none-binary
+ * representations like JSON or XML.
+ *
+ * In practice, this function augments phydat_unit_to_str() with additional
```
s/augments/extends/ ?

haukepetersen commented:
addressed comments

aabadie commented on sys/phydat/phydat_json.c:
```diff
@@ -0,0 +1,99 @@
+#define STATIC_LEN      (14U)
+
+/**
+ * @note    @p buf must be at least 5 byte of size
```
s/byte/bytes/

smlng commented:
needs rebase

haukepetersen commented:
will do

miri64 commented:
Related: #13154

haukepetersen commented:
Although old, this PR should still be useful. The overall form might have some room for future improvement, but merging this as a base seems valid to me.

All deps are merged and this PR is rebased and all open comments are addressed, so let me know when I should squash.

haukepetersen commented:
@aabadie @kaspar030 @miri64 Anyone care to take a final look?

miri64 commented on sys/phydat/phydat_str.c:
```diff
@@ -132,6 +132,18 @@ const char *phydat_unit_to_str(uint8_t unit)
+    switch (unit) {
+        case UNIT_UNDEF:    return "undefined";
+        case UNIT_NONE:     return "none";
+        case UNIT_BOOL:     return "bool";
```
Given that the type for booleans is already defined in JSON, would it make sense to save some bytes here and also map `UNIT_BOOL` this to `"none"` here?

miri64 commented on sys/include/phydat.h:
```diff
@@ -178,6 +179,23 @@ void phydat_dump(phydat_t *data, uint8_t dim);
 
+/**
+ * @brief   Return a string representation for every unit, including
+ *          non-physical units like 'bool' or 'time'
```
This needs to be adapted then of course ;-)

miri64 commented on tests/unittests/tests-phydat/tests-phydat.c:
```diff
@@ -114,10 +177,64 @@ static void test_phydat_fit(void)
+    TEST_ASSERT_EQUAL_STRING("bool", phydat_unit_to_str_verbose(UNIT_BOOL));
+    TEST_ASSERT_EQUAL_STRING("°C", phydat_unit_to_str_verbose(UNIT_TEMP_C));
+    TEST_ASSERT_EQUAL_STRING("Bar", phydat_unit_to_str_verbose(UNIT_BAR));
+    TEST_ASSERT_EQUAL_STRING("ppm", phydat_unit_to_str_verbose(UNIT_PPM));
```
Why are `UNIT_TIME` and `UNIT_DATE` not tested here?

miri64 commented on tests/unittests/tests-phydat/tests-phydat.c:
```diff
@@ -20,6 +35,54 @@
+        .dim  = 3,
+        .dat  = { { 1, 0, 1 }, UNIT_BOOL, 0 },
+        .json = "{\"d\":[true,false,true],\"u\":\"bool\"}"
+    }
```
Same here.

fjmolinas commented:
ping @aabadie!

haukepetersen commented:
As suggested:
- bool values in JSON have now the unit string `none`
- fixed documentation accordingly
- added test cases for date and time units
- added also test cases for all other unit strings while I was at it

haukepetersen commented:
@miri64 would you be ok with me squashing this PR?

kaspar030 commented:
@haukepetersen please squash. @miri64 all yours!

miri64 commented:
@haukepetersen yes please :-)

miri64 commented on sys/phydat/phydat_json.c:
```diff
@@ -0,0 +1,99 @@
+            buf[pos++] = '[';
+        }
+        for (size_t i = 0; i < dim; i++) {
+            if (data->unit != UNIT_BOOL) {
```
Nit-pick: reverse logic. It is easier for the brain to parse to handle the `data->unit == UNIT_BOOL` first (the connection to the logical expression to the called function's name is much closer)

haukepetersen replied:
But as the first case is the more probable case, the present ordering has a higher chance to result in a binary that saves an instruction for this case ;-)

miri64 replied:
I think all our toolchains are smart enough by now to reorder it as we want... But fine... not dying on that hill ;-).

haukepetersen replied:
not quite, as the toolchain can not judge the probability here... The key impact is the CPU: on anything desktop-ish it does indeed not matter as the branch prediction in the CPU will make all paths equally fast by caching them. Also some RIOT platforms (aka m3, m4) do have some form of branch prediction. But the less advanced platforms (avr8, m0) do not support such a thing and hence manual ordering of conditions has an impact, though arguably not a significant one...

haukepetersen commented:
squashed.

@miri64 let me know if you want me to change the condition to change some efficiency for readability...

miri64 commented:
> @miri64 let me know if you want me to change the condition to change some efficiency for readability...

See https://github.com/RIOT-OS/RIOT/pull/7271#discussion_r428014488 I don't think we trade-off efficiency here.
